### PR TITLE
RTC implementation

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements the RTC server. Most of the WebRTC logic (SFU implementation) has been ported from Calls plugin. This was done in a such a way that the plugin can now directly import the `rtc` package and work just like before (see https://github.com/mattermost/mattermost-plugin-calls/pull/40).

Implementation wise I kept the pattern very similar to previous services, relying on message passing when feasible to easily allow for usage over network as well. The next step in fact will be to make this work through the `service` package and enable the `rtcd` service from the plugin.

I added some tests and documentation to hopefully make the overall functioning clearer to understand. These are not comprehensive yet, something to improve in a later refinement.

### CODE REVIEW

File: docs/implementation.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+```mermaid
+flowchart TD
+service-->api[api\nHTTP API Server]
+service-->ws[ws\nWebSocket Server]
+service-->rtc[rtc\nWebRTC SFU]
+service-->store[store\nK/V Store]
+service-->auth[auth\nAuthentication Service]
+auth-->store
+```
```
Comment (cpoile) [inline docs/implementation.md:18]: This is really cool, nicely done.

File: docs/implementation.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+The `auth` packages implements a simple authentication service to register, unregister and authenticate clients.
+
+### `store`
+
+Store provides a basic interface to key-value store. Its main implementation is currently based on [bitcask](https://git.mills.io/prologic/bitcask), a persistent embedded key-value store.
+
+## Sample deployment
+
+```mermaid
+flowchart TB
```
Comment (cpoile) [inline docs/implementation.md:48]: now you're just showing off 😝

File: service/rtc/msg.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/pion/webrtc/v3"
+)
+
+type MessageType int
+
+const (
+	ICEMessage MessageType = iota + 1
```
Comment (cpoile) [inline service/rtc/msg.go:16]: just curious, why +1? Is 0 reserved (or implicitly means nil)?

File: service/service.go [hunk 4]
```diff
@@ -72,17 +79,47 @@ func New(cfg Config, log *mlog.Logger) (*Service, error) {
...
+			switch msg.Type {
+			case ws.OpenMessage:
+				s.log.Debug("connect", mlog.String("connID", msg.ConnID))
+			case ws.CloseMessage:
+				s.log.Debug("disconnect", mlog.String("connID", msg.ConnID))
+			case ws.TextMessage:
+				s.log.Warn("unexpected text message", mlog.String("connID", msg.ConnID))
+			case ws.BinaryMessage:
+				var clientMsg ClientMessage
+				if err := clientMsg.Unpack(msg.Data); err != nil {
```
Comment (cpoile) [inline service/service.go:100]: Is this going to be a hot path? Should we unpack in a go routine? Or is it negligible?

**Review by cpoile — APPROVED** [review]
Really nice, thanks Claudio!

File: service/service.go (hunk 4 above)
Comment (streamer45) [reply service/service.go:100]: Only time (and data) will tell us. At the moment we have a very low rate of events per second, even when lots of users are joining calls. Keeping it simple for now until further optimization is required.

File: service/rtc/msg.go (hunk 3 above)
Comment (streamer45) [reply service/rtc/msg.go:16]: 0 being the default value for a integer variable I tend to avoid using it so that the event has to be explicitly set.

**Pull request merged**